contracts/volt/VoltV2.sol

```diff
@@ -0,0 +1,440 @@
+    mapping(address => mapping(address => uint256)) internal allowances;
+
+    /// @notice Official record of token balances for each account
+    mapping(address => uint256) internal balances;
+
+    /// @notice A record of each accounts delegate
+    mapping(address => address) public delegates;
+
+    /// @notice A checkpoint for marking number of votes from a given block
+    /// @dev  We use uint224 for the votes so we can pack this struct
+    /// into a single storage slot, therefore making transfers etc,
+    /// cheaper for our users, this follows the same pattern as
+    /// OpenZeppelins ERC20Votes which can be found here
+    /// https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/extensions/ERC20Votes.sol
+    struct Checkpoint {
+        uint32 fromBlock;
+        uint224 votes;
+    }
+
+    /// @notice A record of votes checkpoints for each account, by index
+    mapping(address => mapping(uint32 => Checkpoint)) public checkpoints;
+
+    /// @notice The number of checkpoints for each account
+    mapping(address => uint32) public numCheckpoints;
+
+    /// @notice The EIP-712 typehash for the contract's domain
+    bytes32 public constant DOMAIN_TYPEHASH =
+        keccak256(
+            "EIP712Domain(string name,uint256 chainId,address verifyingContract)"
+        );
+
+    /// @notice The EIP-712 typehash for the delegation struct used by the contract
+    bytes32 public constant DELEGATION_TYPEHASH =
+        keccak256("Delegation(address delegatee,uint256 nonce,uint256 expiry)");
+
+    /// @notice The EIP-712 typehash for the permit struct used by the contract
+    bytes32 public constant PERMIT_TYPEHASH =
+        keccak256(
+            "Permit(address owner,address spender,uint256 value,uint256 nonce,uint256 deadline)"
+        );
+
+    /// @notice A record of states for signing / validating signatures
+    mapping(address => uint256) public nonces;
+
+    /// @notice An event thats emitted when an account changes its delegate
+    event DelegateChanged(
+        address indexed delegator,
+        address indexed fromDelegate,
+        address indexed toDelegate
+    );
+
+    /// @notice An event thats emitted when a delegate account's vote balance changes
+    event DelegateVotesChanged(
+        address indexed delegate,
+        uint256 previousBalance,
+        uint256 newBalance
+    );
+
+    /// @notice The standard EIP-20 transfer event
+    event Transfer(address indexed from, address indexed to, uint256 amount);
+
+    /// @notice The standard EIP-20 approval event
+    event Approval(
+        address indexed owner,
+        address indexed spender,
+        uint256 amount
+    );
+
+    constructor(address core) VoltCoreRef(core) {}
+
+    /// @notice Mint new tokens
+    /// @param dst The address of the destination account
+    /// @param amount The number of tokens to be minted
+    function mint(address dst, uint256 amount) external onlyMinter {
```

Comment:
Need an explicit check in this function that the total supply isn't greater than uint224 max after this function is called.

If minting to a destination that hasn't delegated, then we can get a user that has a balance over uint224 max, which means that if they ever try to delegate, they cannot because writeCheckpoint will fail due to the amount of tokens they own being over uint224 max.